File: CreateDummyFiles/Class/ClsAppArg.cs
using System;
using System.Collections.Generic;
using System.IO;
using CmnClsLib.Class;
using CmnClsLib.Module;

// 2026/08/15 Gemini 3.6 Flash (High) Review & Modified

namespace CreateDummyFiles.Class
{
    /// <summary>
    /// アプリケーションのコマンドライン引数を解析し、各種プロパティを設定するクラスです。
    /// </summary>
    public class ClsAppArg
    {
        private ClsLogger _logger;
        private ClsProp _prop;
        private ClsCmmnArgs _cmmnArgs;

        /// <summary>
        /// <see cref="ClsAppArg"/> クラスの新しいインスタンスを初期化します。
        /// </summary>
        /// <param name="logger">ログ出力制御用オブジェクト</param>
        /// <param name="prop">アプリケーションプロパティ設定用オブジェクト</param>
        /// <example>
        /// <code>
        /// var logger = new ClsLogger();
        /// var prop = new ClsProp();
        /// var appArg = new ClsAppArg(logger, prop);
        /// </code>
        /// </example>
        public ClsAppArg(ClsLogger logger, ClsProp prop)
        {
            _logger = logger;
            _prop = prop;
            _cmmnArgs = new(_logger);
            _cmmnArgs.GetModuleInfo(System.Diagnostics.Process.GetCurrentProcess().MainModule?.FileName ?? string.Empty);
            _prop.ExeDir = _cmmnArgs.ExeDir;
            _prop.ExeBaseName = _cmmnArgs.ExeBaseName;
            _prop.Pid = _cmmnArgs.Pid;
        }

        /// <summary>
        /// アプリケーションの設定プロパティを取得または設定します。
        /// </summary>
        /// <example>
        /// <code>
        /// var currentProp = appArg.Prop;
        /// </code>
        /// </example>
        public ClsProp Prop { get => _prop; set => _prop = value; }

        /// <summary>
        /// コマンドライン引数を解析し、アプリケーションプロパティへ設定および検証を行います。
        /// </summary>
        /// <param name="args">コマンドライン引数の配列</param>
        /// <returns>すべての引数が正常に解析・検証された場合は <c>true</c>。無効または不足している引数がある場合は <c>false</c>。</returns>
        /// <example>
        /// <code>
        /// string[] args = ["-d", @"C:\Temp", "-s", "100MB"];
        /// bool isValid = appArg.Parse(args);
        /// </code>
        /// </example>
        public bool Parse(string[] args)
        {
            Dictionary<string, string> namedArguments = MdlArg.GetNamedArgs(args, false);
            _cmmnArgs.NamedArgs = namedArguments;
            bool isValid = _cmmnArgs.GetCommonArgs();
            string tempValue = string.Empty;

            // -----------------------------------------------------------------
            // ClsCmmnParams引数取得：ETC
            // -----------------------------------------------------------------
            _prop.UsageFlag = _cmmnArgs.IsUsage ? ClsProp.USAGE_USAGE : ClsProp.USAGE_NONE;
            _prop.Verbose = _cmmnArgs.Verbose;
            _prop.IsStackTrace = _cmmnArgs.IsStackTrace;

            // -----------------------------------------------------------------
            // Option：
            ReadOnlySpan<string[]> optionKeys =
            [
                ["d", "out"],
                ["f", "date-from"],
                ["t", "date-to"],
                ["n", "name"],
                ["s", "size"]
            ];

            foreach (string key in optionKeys[0])
            {
                if (MdlArg.ContainsKey(namedArguments, key))
                {
                    tempValue = MdlArg.GetValue(namedArguments, key);
                    if (!string.IsNullOrEmpty(tempValue))
                    {
                        _prop.OutputDir = tempValue;
                        break;
                    }
                }
            }

            foreach (string key in optionKeys[1])
            {
                if (MdlArg.ContainsKey(namedArguments, key))
                {
                    tempValue = MdlArg.GetValue(namedArguments, key);
                    if (!string.IsNullOrEmpty(tempValue))
                    {
                        int parsedInt = MdlUtil.ParseInt(tempValue, MdlConst.INT_NULL);
                        if (parsedInt != MdlConst.INT_NULL)
                        {
                            _prop.DateFrom = parsedInt;
                        }
                        break;
                    }
                }
            }

            foreach (string key in optionKeys[2])
            {
                if (MdlArg.ContainsKey(namedArguments, key))
                {
                    tempValue = MdlArg.GetValue(namedArguments, key);
                    if (!string.IsNullOrEmpty(tempValue))
                    {
                        int parsedInt = MdlUtil.ParseInt(tempValue, MdlConst.INT_NULL);
                        if (parsedInt != MdlConst.INT_NULL)
                        {
                            _prop.DateTo = parsedInt;
                        }
                        break;
                    }
                }
            }

            if (_prop.DateFrom > _prop.DateTo)
            {
                _logger.WriteLine(MdlConst.LVL_E, $"INVALID ARGUMENT : -f {_prop.DateFrom} < -t {_prop.DateTo}");
                isValid = false;
            }

            foreach (string key in optionKeys[3])
            {
                if (MdlArg.ContainsKey(namedArguments, key))
                {
                    tempValue = MdlArg.GetValue(namedArguments, key);
                    if (!string.IsNullOrEmpty(tempValue))
                    {
                        _prop.FilenameTemplate = tempValue;
                        break;
                    }
                }
            }

            foreach (string key in optionKeys[4])
            {
                if (MdlArg.ContainsKey(namedArguments, key))
                {
                    tempValue = MdlArg.GetValue(namedArguments, key);
                    if (!string.IsNullOrEmpty(tempValue))
                    {
                        _prop.HumanReadableFileSize = tempValue;
                        break;
                    }
                }
            }

            if (string.IsNullOrEmpty(_prop.OutputDir))
            {
                _logger.WriteLine(MdlConst.LVL_E, $"INVALID ARGUMENT : -d is empty : {_prop.OutputDir}");
                isValid = false;
            }

            try
            {
                _prop.FileSize = _prop.GetFileSize(_prop.HumanReadableFileSize);
            }
            catch (Exception ex)
            {
                _logger.WriteLine(MdlConst.LVL_E, $"INVALID ARGUMENT : -size {_prop.HumanReadableFileSize} ⇒ {ex.Message}");
                isValid = false;
            }

            // -----------------------------------------------------------------
            // 掃除
            // -----------------------------------------------------------------
            namedArguments.Clear();

            return isValid;
        }

        /// <summary>
        /// コマンドラインの各種オプションおよび使用方法（ヘルプメッセージ）をログに出力します。
        /// </summary>
        /// <example>
        /// <code>
        /// appArg.ShowUsage();
        /// </code>
        /// </example>
        public void ShowUsage()
        {
            string dateFromFormatted = MdlDate.GetFormattedDate(_prop.GetDateTimeFromOffset(_prop.DateFrom), "yyyy/MM/dd HH:mm:ss");
            string dateToFormatted = MdlDate.GetFormattedDate(_prop.GetDateTimeFromOffset(_prop.DateTo), "yyyy/MM/dd HH:mm:ss");

            _logger.WriteLine(MdlConst.LVL_NONE, string.Empty);
            _logger.WriteLine(MdlConst.LVL_NONE, $"Usage : {_prop.ExeDir}{Path.DirectorySeparatorChar}{_prop.ExeBaseName}.exe [Option] [Option]...");
            _logger.WriteLine(MdlConst.LVL_NONE, string.Empty);
            _logger.WriteLine(MdlConst.LVL_NONE, "Basic Option：");
            _logger.WriteLine(MdlConst.LVL_NONE, $"   -d|-out path        ：出力先ディレクトリ     （現在値={_prop.OutputDir}");
            _logger.WriteLine(MdlConst.LVL_NONE, $"   -f|-date-from +|-num：何日前からかの日数     （現在値={_prop.DateFrom} ⇒ {dateFromFormatted}");
            _logger.WriteLine(MdlConst.LVL_NONE, $"   -t|-date-to +|-num  ：何日後までかの日数     （現在値={_prop.DateTo} ⇒ {dateToFormatted}");
            _logger.WriteLine(MdlConst.LVL_NONE, $"   -nl-name name       ：ファイル名テンプレート （現在値={_prop.FilenameTemplate}");
            _logger.WriteLine(MdlConst.LVL_NONE, $"   -s|-size size       ：1-1024B|KB|MB|GB       （現在値={_prop.HumanReadableFileSize} ⇒ {_prop.FileSize}Byte");
            _logger.WriteLine(MdlConst.LVL_NONE, string.Empty);
            _logger.WriteLine(MdlConst.LVL_NONE, "Other Option：");
            _logger.WriteLine(MdlConst.LVL_NONE, $"   -v | -vv            ：冗長表示                          （現在値={_prop.Verbose}）");
            _logger.WriteLine(MdlConst.LVL_NONE, $"   -stacktrace         ：例外時STACKTRACE表示フラグ        （現在値={_prop.IsStackTrace}）");
            _logger.WriteLine(MdlConst.LVL_NONE, "   -console mode       ：メッセージ表示 off|stdout|stderr");
            _logger.WriteLine(MdlConst.LVL_NONE, string.Empty);
            _logger.WriteLine(MdlConst.LVL_NONE, "※ファイル名テンプレート書式指定子：%Y、%y、%m、%d、%H、%M、%S、%w、_COMPUTERNAME_、_USERNAME_");
            _logger.WriteLine(MdlConst.LVL_NONE, string.Empty);
            _logger.WriteLine(MdlConst.LVL_NONE, $"Return Code : {MdlConst.LVL_I}:SUCCESS / {MdlConst.LVL_E}:ERROR");
            _logger.WriteLine(MdlConst.LVL_NONE, string.Empty);
        }
    }
}

File: src/main/java/tool/ClsAppArg.java
package tool;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import tool.cmnclslib.cls.ClsCmmnArgs;
import tool.cmnclslib.cls.ClsLogger;
import tool.cmnclslib.mdl.MdlArg;
import tool.cmnclslib.mdl.MdlConst;
import tool.cmnclslib.mdl.MdlDate;
import tool.cmnclslib.mdl.MdlUtil;

/**
 * アプリケーションのコマンドライン引数を解析し、各種プロパティを設定するクラスです。
 */
public class ClsAppArg {

    private static final List<String> OUT_KEYS = List.of("d", "out");
    private static final List<String> FROM_KEYS = List.of("f", "date-from");
    private static final List<String> TO_KEYS = List.of("t", "date-to");
    private static final List<String> NAME_KEYS = List.of("n", "name");
    private static final List<String> SIZE_KEYS = List.of("s", "size");

    private final ClsLogger logger;
    private ClsProp prop;
    private final ClsCmmnArgs cmmnArgs;

    /**
     * {@link ClsAppArg} クラスの新しいインスタンスを初期化します。
     *
     * @param logger ログ出力制御用オブジェクト
     * @param prop   アプリケーションプロパティ設定用オブジェクト
     */
    public ClsAppArg(ClsLogger logger, ClsProp prop) {
        this.logger = Objects.requireNonNull(logger, "logger must not be null");
        this.prop = Objects.requireNonNull(prop, "prop must not be null");
        this.cmmnArgs = new ClsCmmnArgs(this.logger);
        this.cmmnArgs.getModuleInfo();
        this.prop.setExeDir(this.cmmnArgs.getExeDir());
        this.prop.setExeBaseName(this.cmmnArgs.getExeBaseName());
        this.prop.setPid(this.cmmnArgs.getPid());
    }

    /**
     * アプリケーションの設定プロパティを取得します。
     *
     * @return アプリケーション設定プロパティ
     */
    public ClsProp getProp() {
        return prop;
    }

    /**
     * アプリケーションの設定プロパティを設定します。
     *
     * @param prop アプリケーション設定プロパティ
     */
    public void setProp(ClsProp prop) {
        this.prop = Objects.requireNonNull(prop, "prop must not be null");
    }

    /**
     * コマンドライン引数を解析し、アプリケーションプロパティへ設定および検証を行います。
     *
     * @param args コマンドライン引数の配列
     * @return すべての引数が正常に解析・検証された場合は {@code true}、無効または不足している引数がある場合は {@code false}
     */
    public boolean parse(String[] args) {
        Objects.requireNonNull(args, "args must not be null");

        var namedArguments = MdlArg.getNamedArgs(args, false);
        cmmnArgs.setNamedArgs(namedArguments);
        var isValid = cmmnArgs.getCommonArgs();

        // -----------------------------------------------------------------
        // ClsCmmnArgs引数取得：ETC
        // -----------------------------------------------------------------
        prop.setUsageFlag(cmmnArgs.isUsage() ? ClsProp.USAGE_USAGE : ClsProp.USAGE_NONE);
        prop.setVerbose(cmmnArgs.getVerbose());
        prop.setStackTrace(cmmnArgs.isStackTrace());

        // -----------------------------------------------------------------
        // 各オプションの抽出と設定
        // -----------------------------------------------------------------
        findFirstValue(namedArguments, OUT_KEYS).ifPresent(prop::setOutputDir);

        findFirstValue(namedArguments, FROM_KEYS).ifPresent(val -> {
            var parsed = MdlUtil.parseInt(val, MdlConst.INT_NULL);
            if (parsed != MdlConst.INT_NULL) {
                prop.setDateFrom(parsed);
            }
        });

        findFirstValue(namedArguments, TO_KEYS).ifPresent(val -> {
            var parsed = MdlUtil.parseInt(val, MdlConst.INT_NULL);
            if (parsed != MdlConst.INT_NULL) {
                prop.setDateTo(parsed);
            }
        });

        if (prop.getDateFrom() > prop.getDateTo()) {
            logger.writeLine(MdlConst.LVL_E, "INVALID ARGUMENT : -f " + prop.getDateFrom() + " < -t " + prop.getDateTo());
            isValid = false;
        }

        findFirstValue(namedArguments, NAME_KEYS).ifPresent(prop::setFilenameTemplate);

        findFirstValue(namedArguments, SIZE_KEYS).ifPresent(prop::setHumanReadableFileSize);

        if (prop.getOutputDir().isBlank()) {
            if (!cmmnArgs.isUsage()) {
                logger.writeLine(MdlConst.LVL_E, "INVALID ARGUMENT : -d is empty : " + prop.getOutputDir());
            }
            isValid = false;
        }

        try {
            prop.setFileSize(prop.parseFileSize(prop.getHumanReadableFileSize()));
        } catch (IllegalArgumentException ex) {
            logger.writeLine(MdlConst.LVL_E, "INVALID ARGUMENT : -size " + prop.getHumanReadableFileSize() + " \u21D2 " + ex.getMessage());
            isValid = false;
        }

        return isValid;
    }

    /**
     * 指定されたキーリストのうち最初に見つかった非空の値を返します。
     *
     * @param namedArguments 名前付き引数マップ
     * @param keys           検索対象のキーリスト
     * @return 最初に見つかった値を含む {@link Optional}、存在しない場合は {@link Optional#empty()}
     */
    private static Optional<String> findFirstValue(Map<String, String> namedArguments, List<String> keys) {
        for (var key : keys) {
            if (MdlArg.containsKey(namedArguments, key)) {
                var val = MdlArg.getValue(namedArguments, key);
                if (val != null && !val.isBlank()) {
                    return Optional.of(val);
                }
            }
        }
        return Optional.empty();
    }

    /**
     * コマンドラインの各種オプションおよび使用方法（ヘルプメッセージ）をログに出力します。
     */
    public void showUsage() {
        LocalDateTime dateFromTime = prop.getDateTimeFromOffset(prop.getDateFrom());
        LocalDateTime dateToTime = prop.getDateTimeFromOffset(prop.getDateTo());
        String dateFromFormatted = MdlDate.getFormattedDate(dateFromTime, "yyyy/MM/dd HH:mm:ss");
        String dateToFormatted = MdlDate.getFormattedDate(dateToTime, "yyyy/MM/dd HH:mm:ss");

        String progName = prop.getExeBaseName();
        String exeCommand;
        if (progName == null || progName.isBlank() || progName.contains("apiguardian") || progName.contains("surefire")) {
            exeCommand = "java -jar CreateDummyFiles.jar";
        } else if (progName.endsWith(".jar")) {
            exeCommand = "java -jar " + progName;
        } else {
            exeCommand = progName;
        }

        logger.writeLine(MdlConst.LVL_NONE, "");
        logger.writeLine(MdlConst.LVL_NONE, "Usage : " + exeCommand + " [Option] [Option]...");
        logger.writeLine(MdlConst.LVL_NONE, "");
        logger.writeLine(MdlConst.LVL_NONE, "Basic Option：");
        logger.writeLine(MdlConst.LVL_NONE, "   -d|-out path        ：出力先ディレクトリ     （現在値=" + prop.getOutputDir() + "）");
        logger.writeLine(MdlConst.LVL_NONE, "   -f|-date-from +|-num：何日前からかの日数     （現在値=" + prop.getDateFrom() + " \u21D2 " + dateFromFormatted + "）");
        logger.writeLine(MdlConst.LVL_NONE, "   -t|-date-to +|-num  ：何日後までかの日数     （現在値=" + prop.getDateTo() + " \u21D2 " + dateToFormatted + "）");
        logger.writeLine(MdlConst.LVL_NONE, "   -n|-name name       ：ファイル名テンプレート （現在値=" + prop.getFilenameTemplate() + "）");
        logger.writeLine(MdlConst.LVL_NONE, "   -s|-size size       ：1-1024B|KB|MB|GB       （現在値=" + prop.getHumanReadableFileSize() + " \u21D2 " + prop.getFileSize() + "Byte）");
        logger.writeLine(MdlConst.LVL_NONE, "");
        logger.writeLine(MdlConst.LVL_NONE, "Other Option：");
        logger.writeLine(MdlConst.LVL_NONE, "   -v | -vv            ：冗長表示                          （現在値=" + prop.getVerbose() + "）");
        logger.writeLine(MdlConst.LVL_NONE, "   -stacktrace         ：例外時STACKTRACE表示フラグ        （現在値=" + prop.isStackTrace() + "）");
        logger.writeLine(MdlConst.LVL_NONE, "   -console mode       ：メッセージ表示 off|stdout|stderr");
        logger.writeLine(MdlConst.LVL_NONE, "");
        logger.writeLine(MdlConst.LVL_NONE, "※ファイル名テンプレート書式指定子：%Y、%y、%m、%d、%H、%M、%S、%w、_COMPUTERNAME_、_USERNAME_");
        logger.writeLine(MdlConst.LVL_NONE, "");
        logger.writeLine(MdlConst.LVL_NONE, "Return Code : " + MdlConst.LVL_I + ":SUCCESS / " + MdlConst.LVL_E + ":ERROR");
        logger.writeLine(MdlConst.LVL_NONE, "");
    }
}
